rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/transform/TbRenameKeysNode.java

```diff
@@ -0,0 +1,94 @@
+import org.thingsboard.rule.engine.api.util.TbNodeUtils;
+import org.thingsboard.server.common.data.plugin.ComponentType;
+import org.thingsboard.server.common.msg.TbMsg;
+import org.thingsboard.server.common.msg.TbMsgMetaData;
+
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+@Slf4j
+@RuleNode(
+        type = ComponentType.TRANSFORMATION,
+        name = "rename keys",
+        configClazz = TbRenameKeysNodeConfiguration.class,
+        nodeDescription = "Renames msg data or metadata keys to the new key names selected in the key mapping.",
+        nodeDetails = "If the key that is selected in the key mapping is missed in the selected msg source(data or metadata), it will be ignored." +
+                " If the msg data is not a JSON object returns the incoming message as outbound message with <code>Failure</code> chain," +
+                " otherwise returns transformed messages via <code>Success</code> chain",
+        uiResources = {"static/rulenode/rulenode-core-config.js"},
+        configDirective = "tbTransformationNodeRenameKeysConfig",
+        icon = "find_replace"
+)
+public class TbRenameKeysNode implements TbNode {
+
+    TbRenameKeysNodeConfiguration config;
+
+    @Override
+    public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
+        this.config = TbNodeUtils.convert(configuration, TbRenameKeysNodeConfiguration.class);
+    }
+
+    @Override
+    public void onMsg(TbContext ctx, TbMsg msg) throws ExecutionException, InterruptedException, TbNodeException {
+        Map<String, String> renameKeysMapping = config.getRenameKeysMapping();
+        TbMsgMetaData metaData = msg.getMetaData();
+        String data = msg.getData();
+        if (config.isFromMetadata()) {
+            Map<String, String> metaDataMap = metaData.getData();
+            renameKeysMapping.forEach((nameKey, newNameKey) -> {
+                if (metaDataMap.containsKey(nameKey)) {
+                    metaDataMap.put(newNameKey, metaDataMap.get(nameKey));
+                    metaDataMap.remove(nameKey);
+                }
+            });
+            metaData = new TbMsgMetaData(metaDataMap);
+        } else {
+            JsonNode dataNode = JacksonUtil.toJsonNode(msg.getData());
+            if (dataNode.isObject()) {
+                ObjectNode msgData = (ObjectNode) dataNode;
+                renameKeysMapping.forEach((nameKey, newNameKey) -> {
+                    if (msgData.has(nameKey)) {
+                        msgData.set(newNameKey, msgData.get(nameKey));
+                        msgData.remove(nameKey);
+                    }
+                });
+                data = JacksonUtil.toString(msgData);
+            } else {
+                ctx.tellFailure(msg, new RuntimeException("Msg data is not a JSON Object!"));
+                return;
+            }
+        }
+        ctx.tellSuccess(TbMsg.transformMsg(msg, msg.getType(), msg.getOriginator(), metaData, data));
```

Comment:
No need to transform the message if nothing is changed. This is a simple optimization that will save efforts on copying the message if nothing changed.